### ISSUE TITLE
With -W I get a spurious 'will not stay shared' message

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9205]
Migrated from [rt.perl.org#50160](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=50160) (status was 'resolved')

Searchable as RT50160$

**Comment by p5pRT** [comment]
### From ericp@ActiveState.com

#### Created by ericp@activestate.com




Given this code&#8203;:
\# \<\<\<\<\<\<

sub xeval \{
&nbsp;    my @&#8203;dbline = \('$a\{$b\[3\] \+ 4\}'\, '\(\($b \+ 3\) \* 4\)'\);
&nbsp;    my @&#8203;ret;
&nbsp;    for my $\_sourceString \(@&#8203;dbline\) \{
&nbsp;        my $sourceString = $\_sourceString;
&nbsp;        my %opposites =  \('\{' => '\}'\, '\[' => '\]'\, '\(' => '\)'\, '\<' => '>'\, \);
&nbsp;        $sourceString =~ s/\\bqr\(\[\\\{\\\[\\\(\\\<\]\)\.\*?\(??\{quotemeta $opposites\{$1\}\}\)//g;
&nbsp;        push @&#8203;ret\, $sourceString;
&nbsp;    \}
&nbsp;    return @&#8203;ret;
\}

my @&#8203;res = eval \{ xeval\(\) \};
print join\("\, "\, @&#8203;res\)\, "\\n";

\# >>>>>>>

Running it with \-W with perl v5\.8\.8 acts as expected\, but with Perl  v5\.10\.0 I
get this behavior&#8203;:

c&#8203;:> perl \-W repro\.pl
Variable "%opposites" will not stay shared at \(re\_eval 1\) line 1\.
$a\{$b\[3\] \+ 4\}\, \(\($b \+ 3\) \* 4\)

c&#8203;:>

The warning message is unexpected\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.10.0:

Configured by SYSTEM at Thu Nov 22 14:37:24 2007.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=MSWin32, osvers=5.00, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -GF -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -DUSE_SITECUSTOMIZE -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='12.00.8804', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"D:\apps\perl-5.10b\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl510.lib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"D:\apps\perl-5.10b\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    perl-current@32448


@INC for perl 5.10.0:
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -GF -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -DUSE_SITECUSTOMIZE -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='12.00.8804', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"D:\apps\perl-5.10b\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl510.lib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"D:\apps\perl-5.10b\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    perl-current@32448


@INC for perl 5.10.0:
    D:/apps/perl-5.10b/lib
    D:/apps/perl-5.10b/site/lib
    .


Environment for perl 5.10.0:
    HOME=c:\Users\ericp
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=D:\apps\perl-5.10b\bin;C:\PHP;c:\ruby\bin;c:\apps\svn\bin;C:\Python25;C:\Perl\bin;c:\bin;c:\Users\ericp\bin;c:\msys\bin;C:\apps\p4;C:\TclDevKit\bin;C:\Tcl\bin;C:\Program Files\ActiveState Komodo IDE 4.2;C:\Program Files\ActiveStatePerl Dev Kit 7.0\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Program Files\Microsoft SQL Server\90\Tools\binn;C:\Program Files\MySQL\MySQL Server 5.0\bin;C:\Program Files\QuickTime\QTSystem;C:\Program Files\Microsoft SQL Server\90\Tools\binn\
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Jan 23\, 2008 at 11&#8203;:10&#8203;:37AM \-0800\, Eric Promislow wrote&#8203;:
> Given this code&#8203;:
> \# \<\<\<\<\<\<
> 
> sub xeval \{
>     my @&#8203;dbline = \('$a\{$b\[3\] \+ 4\}'\, '\(\($b \+ 3\) \* 4\)'\);
>     my @&#8203;ret;
>     for my $\_sourceString \(@&#8203;dbline\) \{
>         my $sourceString = $\_sourceString;
>         my %opposites =  \('\{' => '\}'\, '\[' => '\]'\, '\(' => '\)'\, '\<' => '>'\, \);
>         $sourceString =~ s/\\bqr\(\[\\\{\\\[\\\(\\\<\]\)\.\*?\(??\{quotemeta $opposites\{$1\}\}\)//g;
>         push @&#8203;ret\, $sourceString;
>     \}
>     return @&#8203;ret;
> \}
> 
> my @&#8203;res = eval \{ xeval\(\) \};
> print join\("\, "\, @&#8203;res\)\, "\\n";
> 
> \# >>>>>>>
> 
> Running it with \-W with perl v5\.8\.8 acts as expected\, but with Perl  v5\.10\.0 I
> get this behavior&#8203;:
> 
> c&#8203;:> perl \-W repro\.pl
> Variable "%opposites" will not stay shared at \(re\_eval 1\) line 1\.
> $a\{$b\[3\] \+ 4\}\, \(\($b \+ 3\) \* 4\)
> 
> c&#8203;:>
> 
> The warning message is unexpected\.

The warning message is correct\. Consider the following code&#8203;:

&nbsp;    sub f \{
&nbsp;    \(my $n\) = @&#8203;\_;
&nbsp;    "a" =~ /a\(??\{ print "n=$n\\n"; "x" \}\)/;
&nbsp;    \}

&nbsp;    f\(1\);
&nbsp;    f\(2\);

Most people's expectations are that it should output

&nbsp;    n=1
&nbsp;    n=2

but the actual output is

&nbsp;    Variable "$n" will not stay shared at \(re\_eval 1\) line 1\.
&nbsp;    n=1
&nbsp;    n=1


So this is really a bug in the way lexical variables are currently handled
within \(?\{\.\.\.\}\) and \(??\{\.\.\.\}\) RE constructs\. This bug may get fixed in 5\.12\.
In the meantime\, it is better to only use package variables within them\.


\-\- 
"There's something wrong with our bloody ships today\, Chatfield\."
&nbsp;    \-\- Admiral Beatty at the Battle of Jutland\, 31st May 1916\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
